### PR TITLE
[LDM] Speed optimization on repetitive data

### PR DESCRIPTION
**Pull request by terrelln**
LDM does especially poorly on repetitive data when that data's hash happens
to have `(hash & stopMask) == 0`. Either because the `stopMask == 0` or
random chance. Optimize this case by skipping over repetitive patterns.
The detection is very simplistic, but should catch most of the offending
cases.

```
head -c 1G /dev/zero | perf stat -- ./zstd -1 -o /dev/null -v --zstd=ldmHashRateLog=1 --long
      21.187881087 seconds time elapsed

head -c 1G /dev/zero | perf stat -- ./zstd -1 -o /dev/null -v --zstd=ldmHashRateLog=1 --long
       1.208744521 seconds time elapsed
```

With this change, I see 1-5% ratio improvements for very compressible data,
like in `--patch-from` mode, because we skip adding the tails of very long
matches to our tables, which leaves room for other candidates.

It is approximately neutral for non `--patch-from` mode, since extremely long
matches aren't very common, so this corner case rarely gets triggered.

### CODE REVIEW

File: lib/compress/zstd_ldm.c [hunk 1]
```diff
@@ -57,6 +57,33 @@ static void ZSTD_ldm_gear_init(ldmRollingHashState_t* state, ldmParams_t const*
     }
 }
 
+/** ZSTD_ldm_gear_reset()
```
Comment (Cyan4973) [inline lib/compress/zstd_ldm.c:60]: 👍

**Comment by Cyan4973** [comment]
So repetitive data is detected whenever an overlapping match happens ?
In which case, the match length is unconditionally skipped, irrespective of the produced hash value (wether `(hash & stopMask) == 0` or not) ?

To be fair, I was expecting this policy to be already in place for any match.
While I would expected such policy to scan faster, I presume it would also cost too much compression ratio ?

**Comment by terrelln** [comment]
> So repetitive data is detected whenever an overlapping match happens ?
> In which case, the match length is unconditionally skipped

Exactly.

> irrespective of the produced hash value (wether (hash & stopMask) == 0 or not) ?

If it is a match, then it was inserted, so we actually know that `(hash & stopMask) == 0`.

> To be fair, I was expecting this policy to be already in place for any match.
> While I would expected such policy to scan faster, I presume it would also cost too much compression ratio ?

LDM has always inserted every position. However, switching to gear-hash made this scenario much worse. We used to use the same hash for the insertion-mask and the table-index. But with gear-hash we use `XXH64()` for the table-index. That means we have to re-hash every position that matches. It isn't a big deal when matches are somewhat infrequent, but repetitive bytes that happened to match can be a bad corner case.

**Review by Cyan4973 — APPROVED** [review]

File: lib/compress/zstd_ldm.c [hunk 2]
```diff
@@ -447,6 +467,17 @@ static size_t ZSTD_ldm_generateSequences_internal(
...
 
             anchor = split + forwardMatchLength;
+
+            /* If there is an overlapping match skip to the end of it.
+             * No need to insert every position into our table. This
+             * is especially slow because position of a long overlapping match
+             * is inserted, every position is inserted.
+             */
+            if (mLength > (size_t)offset) {
+                ip = anchor;
```
Comment (mpu) [inline lib/compress/zstd_ldm.c]: I had a bit of trouble thinking about this because `ip` used to be constant in this loop.

It probably does not matter much but I think this may skip a bit too much. I would find the code easier to reason about if it skipped just the right amount: `ip = anchor - hashed;`

File: lib/compress/zstd_ldm.c [hunk 3]
```diff
@@ -447,6 +467,17 @@ static size_t ZSTD_ldm_generateSequences_internal(
             ZSTD_ldm_insertEntry(ldmState, hash, newEntry, *params);
 
             anchor = split + forwardMatchLength;
+
+            /* If there is an overlapping match skip to the end of it.
+             * No need to insert every position into our table. This
+             * is especially slow because position of a long overlapping match
+             * is inserted, every position is inserted.
+             */
+            if (mLength > (size_t)offset) {
```
Comment (mpu) [inline lib/compress/zstd_ldm.c]: If I understand correctly, we want to save work in the `while (ip < ilimit)` loop.

If that is the case, I would instead detect the degenerate case using `anchor > ip + hashed`, then reset `ip = anchor - hashed` and break instead of continue. It makes it a bit clearer where the savings are expected and also makes it easy to see the relation with the `ip` reset and the increment line 483.

**Review by mpu — COMMENTED** [review]
My two cents!

File: lib/compress/zstd_ldm.c (hunk 3 above)
Comment (terrelln) [reply lib/compress/zstd_ldm.c]: Thats a good point! I actually intended to continue the `while (ip < ilimit)` loop, not the inner loop, so this is a logic bug. What you're saying is what I was intending to write. Thanks for the review!

Comment (terrelln) [reply lib/compress/zstd_ldm.c]: That isn't quite the same logic, since `mLength > offset` detects overlapping matches, whereas `anchor > ip + hashed` detects long matches.

However, I've found that `anchor > ip + hashed` improves compression ratio by a reasonable amount in some `--patch-from` use cases that I tested, where the compression ratio is very high. So it is probably a reasonable choice.

**Comment by terrelln** [comment]
I've updated to the logic suggested by @mpu, but I have to do some more in-depth performance testing, since this covers more than just the overlapping-match edge case. I think it will be positive.
